src/sage/stats/distributions/discrete_gaussian_lattice.py

```diff
@@ -222,51 +229,214 @@ def _normalisation_factor_zz(self, tau=3):
+
+            sage: M = Matrix(ZZ, [[1, 3, 0], [-2, 5, 1]])
+            sage: D = DGL(M, 3)
+            sage: D._normalisation_factor_zz()
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: Basis must be a square matrix for now.
+
+            sage: D = DGL(ZZ^3, c=(1/2, 0, 0))
+            sage: D._normalisation_factor_zz()
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: Lattice must contain 0 for now.
+
+            sage: D = DGL(Matrix(3, 3, 1/2))
+            sage: D._normalisation_factor_zz()
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: Lattice must be integral for now.
         """
-        if self.B != identity_matrix(ZZ, self.B.nrows()):
-            raise NotImplementedError("This function is only implemented when B is an identity matrix.")
 
-        f = self.f
-        n = self.B.ncols()
+        # If σ > 1:
+        # We use the Fourier transform g(t) of f(x) = exp(-k^2 / 2σ^2), but
+        # taking the norm of vector t^2 as input, and with norm_factor factored.
+        # If σ ≤ 1:
+        # The formula in docstring converges quickly since it has -1 / σ^2 in
+        # the exponent
+        def f_or_hat(x):
+            # Fun fact: If you remove this R() and delay the call to return,
+            # It might give an error due to precision error. For example,
+            # RR(1 + 100 * exp(-5.0 * pi^2)) == 0
+
+            if sigma > 1:
+                return R(exp(-pi**2 * (2 * sigma**2) * x))
+
+            return R(exp(-x / (2 * sigma**2)))
+
+        if not self.is_spherical:
+            # TODO: This is only a poor approximation placeholder.
+            # It should be easy to implement, since the Fourier transform
+            # is essentially the same, but I can't figure out how to
+            # tweak the `.qfrep` call below correctly.
+            from warnings import warn
+            warn("Note: `_normalisation_factor_zz` has not been properly "
+                 "implemented for non-spherical distributions.")
+            import itertools
+            from sage.functions.log import log
+            basis = self.B.LLL()
+            base = vector(ZZ, [v.round() for v in basis.solve_left(self._c)])
+            BOUND = max(1, (self._RR(log(10**4, self.n)).ceil() - 1) // 2)
+            if BOUND > 10:
+                BOUND = 10
+            coords = itertools.product(range(-BOUND, BOUND + 1), repeat=self.n)
+            return sum(self.f((vector(u) + base) * self.B) for u in coords)
+
+        if self.B.nrows() != self.B.ncols():
+            raise NotImplementedError("Basis must be a square matrix for now.")
+
+        if self.is_spherical and not self._c_in_lattice:
+            raise NotImplementedError("Lattice must contain 0 for now.")
+
+        if self.B.base_ring() != ZZ:
+            raise NotImplementedError("Lattice must be integral for now.")
```

Comment:
```suggestion
            raise NotImplementedError("only implemented for the basis being a square matrix")

        if self.is_spherical and not self._c_in_lattice:
            raise NotImplementedError("only implemented when the lattice contains 0")

        if self.B.base_ring() != ZZ:
            raise NotImplementedError("only implemented for integral lattices")
```
We also follow Python's convention for error messages (lower case without a period/full-stop).

Reply:
Ooo I didn't know it's a Python convention, but I learned that in the past few months